gtsam/hybrid/HybridSmoother.cpp

```diff
@@ -0,0 +1,112 @@
+/* ----------------------------------------------------------------------------
+
+ * GTSAM Copyright 2010, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file    HybridSmoother.cpp
+ * @brief   An incremental smoother for hybrid factor graphs
+ * @author  Varun Agrawal
+ * @date    October 2022
+ */
+
+#include <gtsam/hybrid/HybridSmoother.h>
+
+#include <algorithm>
+#include <unordered_set>
+
+namespace gtsam {
+
+/* ************************************************************************* */
+void HybridSmoother::update(HybridGaussianFactorGraph graph,
+                            const Ordering &ordering,
+                            boost::optional<size_t> maxNrLeaves) {
+  // Add the necessary conditionals from the previous timestep(s).
+  std::tie(graph, hybridBayesNet_) =
+      addConditionals(graph, hybridBayesNet_, ordering);
+
+  // Eliminate.
+  auto bayesNetFragment = graph.eliminateSequential(ordering);
+
+  /// Prune
+  if (maxNrLeaves) {
+    // `pruneBayesNet` sets the leaves with 0 in discreteFactor to nullptr in
+    // all the conditionals with the same keys in bayesNetFragment.
+    HybridBayesNet prunedBayesNetFragment =
+        bayesNetFragment->prune(*maxNrLeaves);
+    // Set the bayes net fragment to the pruned version
+    bayesNetFragment =
+        boost::make_shared<HybridBayesNet>(prunedBayesNetFragment);
+  }
+
+  // Add the partial bayes net to the posterior bayes net.
+  hybridBayesNet_.push_back<HybridBayesNet>(*bayesNetFragment);
+}
+
+/* ************************************************************************* */
+std::pair<HybridGaussianFactorGraph, HybridBayesNet>
+HybridSmoother::addConditionals(const HybridGaussianFactorGraph &originalGraph,
+                                const HybridBayesNet &originalHybridBayesNet,
+                                const Ordering &ordering) const {
+  HybridGaussianFactorGraph graph(originalGraph);
+  HybridBayesNet hybridBayesNet(originalHybridBayesNet);
+
+  // If we are not at the first iteration, means we have conditionals to add.
+  if (!hybridBayesNet.empty()) {
+    // We add all relevant conditional mixtures on the last continuous variable
+    // in the previous `hybridBayesNet` to the graph
+
+    // Conditionals to remove from the bayes net
+    // since the conditional will be updated.
+    std::vector<HybridConditional::shared_ptr> conditionals_to_erase;
+
+    // New conditionals to add to the graph
+    gtsam::HybridBayesNet newConditionals;
+
+    // NOTE(Varun) Using a for-range loop doesn't work since some of the
+    // conditionals are invalid pointers
```

Comment:
Interesting, I don't think this is possible in a Hybrid Bayes Net?

Reply:
It is! I was able to get a simple example up and running with this.

Reply:
I have no other explanation why a for-range loop doesn't work, unfortunately...